Blueprint Model

H0R5E opened the pull request:
Hi @ryancoe and @ssolson,

I just wanted to show you the early version of the blueprint model before I have to finish for the day. Please focus your attention initially on the top level `experimental\RM3` directory. In there is a partial implementation of the RM3 model using the blueprint structure discussed in #140 and an example script (`example.m`) that follows a similar approach to the `WaveBot_caseA.m` example.

In many ways this also (almost accidentally) looks a lot like #126, which was not my initial intention, but it kind of just came out that way. I ended up defining three main abstracts that are added to the Blueprint class that communicate with intermediate data which ends up being stored in the Device class. All of the abstracts that are used to define the classes are in the `+WecOptTool\+experimental package`, which define a set of interfaces for the user to reimplement to build their blueprint (in the `+blocks` sub-pacakage), an example of reusable classes (in the `+blocks\+geometry` sub-package) and other classes not necessarily meant to be user facing (in the `+shared` sub-package). The classes and methods are a little different to what I proposed in #140 and I will build a new diagram to show the relationships tomorrow.

I don't want to go too much further if this is not what you want, so let me know if it looks like it's going the right way.

Cheers,

Mat

EDIT: A list of must have things TODO to finish (in no particular order):

- [x] Define general structure
- [x] Add multiple sea states
- [x] Add sea state manipulation
- [x] Disconnect Blueprint and Device abstracts
- [x] Add serialisation / de-serialisation
- [x] Expand parameter passing
- [x] Create data type classes
- [x] Finish RM3 port
- [x] Replace WecOptTool with WecOptLib.experimental package
- [x] Remove WecOptLib.model and WecOptLib.nemoh directories
- [x] Update existing tests as required following restructure
- [x] Add minimal inline documentation (at least GPL headers)
- [x] Update example and other online documentation
- [x] Update README

A list of nice to haves but which won't block merging:
-  Finish WaveBot model
-  Finish port of features from #140 
-  Restructure tests
-  Complete inline documentation

Closes #70 
Closes #126

ryancoe commented on toolbox/+WecOptLib/+experimental/+blocks/Blueprint.m:
```diff
@@ -0,0 +1,96 @@
+            
+        end
+        
+        function devices = makeDevices(obj, geomTypes,         ...
```
I'm guilty of it too, but we should maybe use something like `designParams` instead of `geomParams`, as this could include parameters to control PTO gearing, operational limits, etc.

Also, will eventually want to add a `controlParams` argument to handle, e.g., PTO force limitations.

H0R5E replied:
The first arguments are always passed to the Geometry class, so I think make sense for the names to be related to that. We can add additional arguments as required but until it's clearer where parameters like PTO gearing are applied its hard to imagine how is the best way to call the setup. I'll need to put in `controlParams` to finish the RM3 example with PS, but I wonder if it's more useful to be used in the Device class?

H0R5E replied:
Also, perhaps the PTO gearing is like parameters to the model? I'm sure we can provide some kind of generalisation for passing parameters to all of the components.

ryancoe replied:
All sounds good. Just need to remember to address this later.

ryancoe commented on toolbox/+WecOptLib/+experimental/+blocks/Blueprint.m:
```diff
@@ -0,0 +1,96 @@
+                geomParam = geomParams{i};
+                
+                Geometry = obj.Geometries.(geomType);
+                geometry = Geometry();
```
What is this line doing?

H0R5E replied:
Do you mean 77? It's instantiating the Geometry class. The user just hooks up classes and we need to turn them into objects before they can be used.

ryancoe replied:
I did mean 77, thanks.

However, curious about your coding style of creating a variable to point to an element in a cell array. Why not just reference the element directly?

H0R5E replied:
Do you mean doing this?
```
obj.Geometries.(geomTypes{i});
```

I guess I personally find it more readable if there are less embedded brackets, but there is no need to do this step first.

ryancoe commented on experimental/RM3/ComplexCongugate.m:
```diff
@@ -0,0 +1,18 @@
+
+            % Maximum absorbed power
+            % Note: Re{Zi} = Radiation Damping Coeffcient
+            performance = abs(motion.F0) .^ 2 ./ (8 * real(motion.Zi));
```
Not sure if you are planning to do this eventually and just haven't gotten their yet, but we will want to

1. Make it more clear what `performance` is (frequency dependent power in this case)
2. Solve for the device state, etc. and return these as well as discussed in #122 (perhaps with a class as done in  #140 with `SimResults`)

H0R5E replied:
Yeah, I was just doing something quick and dirty here and my intention is to merge #140 into this. I think something similar should be applied to the `motion` struct as well.

ryancoe commented on toolbox/+WecOptLib/+experimental/+blocks/Blueprint.m:
```diff
@@ -0,0 +1,96 @@
+classdef (Abstract) Blueprint
```
I'm sure you were getting there, but even some brief inline docs would help me understanding this structure.

ryancoe commented on toolbox/+WecOptLib/+experimental/+blocks/Model.m:
```diff
@@ -0,0 +1,8 @@
+classdef (Abstract) Model
```
Model may be to ambiguous of a name

H0R5E replied:
Yeah, I went around the houses with this one and kept coming back to model. This is the model that generates the equations of motion, fundamentally, and I couldn't think of anything better.

ryancoe commented on toolbox/+WecOptLib/+experimental/+shared/Device.m:
```diff
@@ -0,0 +1,41 @@
+classdef Device < handle
```
`Device` may be too ambiguous of a name (maybe `WecDevice`?)

H0R5E replied:
Maybe just WEC?

ryancoe commented on experimental/RM3/RM3.m:
```diff
@@ -0,0 +1,17 @@
+    
+    properties
+        
+        Geometries = struct(                                            ...
```
Does this syntax mean you're setting the values of these properties automatically? Or is the defining the acceptable values/forms of the properties?

H0R5E replied:
The user decides which geometries and controllers they want available in their blueprint by adding structs to these properties with the names and classes they want to use. There is only one model for a blueprint, so only a single Model subclass is added.

MATLAB doesn't have enough introspection to make some sort of plugin architecture easy to do and it would not help us when we want to reuse an existing class, such as the `WecOptLib.experimental.blocks.geometry.ExistingNEMOH` class used in `RM3`.

ryancoe commented on toolbox/+WecOptLib/+experimental/+blocks/Model.m:
```diff
@@ -0,0 +1,8 @@
+classdef (Abstract) Model
+    
+    methods (Abstract)
+       static = getStatic(obj, hydro)
+       dynamic = getDynamic(obj, static, hydro, S)
```
don't like the variable name `static` (maybe `hydroStatic`?)

same with `dynamic`

H0R5E replied:
Yeah, happy to take suggestion here, but the idea is that `getStatic` is for parts of the model that do not vary by seastate and `getDynamic` is for the parts of the model that do.

ssolson commented:
I only looked over the code to see the general structure. Looking at the example the workflow does not seem too split up to me. Further looking at the functions changed here is there specifically a part where the workflow seems too broken up? Generally speaking this seems like an improvement in clarity. I like the "blueprint" class as a name as we are truly trying to build/ optimize a device.  Again my current take is that this is similar to the current implementation but better.

H0R5E commented:
OK, thanks for the feedback. You've probably gone further than I was anticipating because this needs a lot more documentation throughout. I agree that having to skip files is a little annoying, and partly that's due MATLABs requirement of only one class per file, but I think that the abstracting/superclass concept will become increasing useful as we find more commonalities and reusable parts of the code. If we can standardise the motion struct, for instance, to always have the forces and impedance matrices defined, then reuse of the simple controllers seems achievable.

I need to think about the documentation (particularly in the abstracts), but I'll try and flesh this out into a full example now, taking some of your comments on board.

ryancoe commented:
> I only looked over the code to see the general structure. Looking at the example the workflow does not seem too split up to me. Further looking at the functions changed here is there specifically a part where the workflow seems too broken up?

It is mostly in trying to dig into the functionality that I feel like things are broken up. It is at least partially a result of, as @H0R5E points out, MATLAB's rule against multiple classes in a file. For instance, I try to figure out how these two lines in `WecOptTool/experimental/RM3/example.m` work:

```matlab
blueprint = RM3();
devices = blueprint.makeDevices(geomMode, geomParams, controllers)
```

so I open up `WecOptTool/experimental/RM3/RM3.m` and see that it inherits from `WecOptLib.experimental.blocks.Blueprint`, so I open up that up and see that it basically points to or imports from other classes, in other files... I think this is mostly just the way it needs to be to have the code be both OOP and DRY, but if there's a way to reduce the numbers of files I would be grateful

H0R5E commented:
OK, so I'm going to stop again because I'm concerned that with encapsulation we are likely to generate lots of files as we break the problem down (for instance the concrete versions of the Mesher and Solver classes would be used inside a Hydrodynamics subclass for the parametric case to allow reuse of those components), so maybe encapsulation is not what you are after.

Here the user has a fixed workflow to follow `Geometry -> Model -> Controller` using classes defined to accept an understood output from each, so Geometry creates a hydro struct, Model creates a motion struct while Controller will create a performance struct. How that data gets passed between the classes they make shouldn't be a concern of the user, IMO.

The approach also favours reusability  as it allows a user to make up their problem from a library of existing classes that can work on the intermediate outputs.

It's hard to think of another way to do this. If we were to try and drop classes for functions, somehow, to provide the same kind of reusability we would generate lots of files containing functions (so still jumping around). Otherwise you generate very long functions (or lots of sub-functions), but then you are back to writing every WEC bespoke or copying and pasting code.

Is there some other piece of software that you think solves this problem in a nice way? If you can find me an example, then I'm sure we can recreate it.

ryancoe commented:
@H0R5E - I don't know of another piece of software that's found a solution to this. I honestly don't know that many third-party toolboxes for MATLAB that use OOP.

I don't think that the large number of files is a show-stopper. In fact, as the code grows (i.e., we have additional devices, the number of files and lines of code will grow slower with this approach because it is more DRY. My comment was more just a question about whether we can _**tweak**_ this approach to reduce the number of files / improve readability. I don't think you should stop. I think you should complete this out.

H0R5E commented:
OK fair enough. I just wanted to make sure that it was clear where this path is leading! I'll keep an eye out for more elegant approaches as I go.

H0R5E commented:
> I honestly don't know that many third-party toolboxes for MATLAB that use OOP.

@ryancoe I wasn't necessarily looking for solutions to organising the objects, more if there is any sort of problem set up approach (in MATLAB or otherwise) that is more intuitive.

H0R5E commented:
Right, I have changed the implementation a bit, to allow use of private functions in the blueprints and to allow us to generate new types of blueprints and devices in the future so that they can be extended without everything breaking.

To give some background regarding this, there are two design patterns being used. On the user side, they are being asked to implement a [strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern) which enables "selecting an algorithm at runtime." In other words, the idea of a strategy pattern is to easily change parts of the algorithms within a program by encapsulating the parts that vary. Normally, this would all be implemented using abstract classes, but I've chosen to use function callbacks now to make it a bit easier to implement. The most critical part is making sure the data types being produced by the callbacks is correct, and that is the next thing I will add to this.

For future development, I have also updated the Blueprint class using the [abstract factory pattern](https://en.wikipedia.org/wiki/Abstract_factory_pattern). Blueprint was always a factory class (as it created Devices) but now both Blueprint and Device are abstract, with the default implementations called DefaultBlueprint and DefaultDevice. What's nice is that because these are derived from abstracts we can generate new versions of the Blueprint and Device (say if we want to add more stages to the problem), but the calling signatures will remain the same, i.e. the example code will not need to change significantly.

Here is a new class diagram which shows implementation of the two patterns:

![WecOptTool (2)](https://user-images.githubusercontent.com/5088535/82763197-cb38b780-9dfd-11ea-909a-938783bfce9f.png)

Finally, I have added some documentation to the parts I think are more solid and I added an example with an optimisation in it (example_optim.m) in the experimental/RM3 folder.

ryancoe commented:
@H0R5E - This looks pretty neat. I have to say that I will need a bit more time (and maybe live walkthrough from you) to more fully understand some of the concepts and structures, but I think I understand the basics. Is the next step to merge #140 into this?

H0R5E commented:
I have added a TODO list in the opening comment.

H0R5E commented:
> Is the next step to merge #140 into this?

I think, given you are actively working with #140 at the moment, I might just cherry pick a few things from it for now, and try and get some sort of final draft for moving the existing code and RM3 example. I'm worried if we both try to use this branch for different purposes that we might clash somewhere.

ryancoe commented:
What do you mean by replace WecOptTool and remove WecOptLib? I do have to say I’m often unsure which directory to look in for things.

H0R5E commented:
I think we need to chat about it (hence the question marks) but my impression is that this work will replace the content of the WecOptTool package (for sure) and that (possibly) we could remove the WecOptLib package - or at least we would aim to phase it out.

The advantage of keeping them separate is that there is a clearer distinction of 'this is for the user' vs 'this is for the implementation (you don't need to see)', which would be harder to distinguish in one unified namespace.

ryancoe commented:
From our chat today, here's my updated understanding:

 - All the code that the user would work on directly to make their own device is demonstrated in `experimental/RM3`
 - The example device `experimental/RM3/RM3` uses function handles (callbacks) that point to functions written inside of `experimental/RM3/RM3.m`. This allows the methods of `WecOptTool/toolbox/+WecOptLib/+experimental/DefaultBlueprint` to call these functions in a dynamic way without the use of switches. It also means that user defined devices can minimize OOP and use functions instead.
 - The abstract classes in `WecOptTool/toolbox/+WecOptLib/+experimental` (`DefaultBlueprint` and `DefaultDevice` are abstracts of those in `WecOptTool/toolbox/+WecOptLib/+experimental/+base`. The reason for doing this is to try to make creating new blueprint or device classes easier.

H0R5E commented:
Thanks Ryan.

We also discussed flexibility around argument passing to the different stages of the algorithm (like we pass GeoParams in the current example). I think there are two possible approaches to extending this functionality to the other stages and I'll outline them so we can make a decision:

1. The signature of makeDevices becomes `makeDevices(geomType, geomParams, controllerType, controlParams, modelParams)` and the user has to pass a place holder if there model doesn't accept parameters for certain configurations. All callbacks will need to accept at least one parameter argument also, even if it's just a dummy one.
2. We use named arguments for the parameters, i.e. `makeDevices(geomType, controllerType, 'geomParams', geomParams, 'controlParams', controlParams', 'modelParams', modelParams)`. This would allow parameters not to be passed when not used and the callbacks can just accept the expected number of inputs.

To me, I think option 2 is the most natural, as there may be some controllers that take arguments, so not and the same with the other stages.

I also want to say that I'm fine with parking the abstract Blueprint and Device classes until we need them. I think you are right that the "Default" part doesn't make any sense until we have more than one anyway and if my use case comes around where we want to offer legacy support we can always bring back the abstract and rename the original Blueprint as LegacyBlueprint or something.

H0R5E commented on experimental/RM3/RM3.m:
```diff
@@ -0,0 +1,219 @@
+    
+    properties
+        
+        geometryCallbacks = struct(                                     ...
```
@ryancoe, do you thing "handle" is a better term to use here? So like `geometryHandles` might be more a MATLABy way of describing what goes here? (or `geometryFunctionHandles` `geometryFHandles`?)

ssolson replied:
I think Callbacks feels contrived. Why not pointer? Handle is okay too but is MATLAB specific.

H0R5E replied:
"Contrived"? Harsh! Anyway, the program is written in MATLAB, so making it MATLAB specific is not a bad thing, IMHO. Now I'm wondering if just appending "Function" is the most obvious kind of naming, however.

ssolson replied:
Apologies Mat. Not meant to be harsh! When I asked what "CallBack" meant you and Ryan both said pointer so that seems like the easiest name. Then secondly Handle because of MATLAB. I also still dream that this program will not always be in MATLAB.

H0R5E commented:
OK, so I have added a basic implementation of all the data types and simulation of multiple sea states. The commit message is as follows:

> The four defined data types are added to this commit (SeaState, Hydro, Motion and Performance), which are based upon the abstract Data class. The type classes enforce inclusion of certain parameters when initialized (from a struct) and will store any additional parameters given. They can also be initialized from the struct array. As in SeaState, the type classes can also be used for data preparation, as required. The required user defined functions must also return one of
> these types.
> 
> Multiple sea-states are now solved by DefaultDevice.simulate and the Motion and Performance object created for each are stored as an array (in the motions and performances properties). The user can now add a optional additional callback to DefaultBlueprint that will take the SeaState object and all of the results in DefaultDevice and any output from this will be added to the aggregation property of DefaultDevice. This could also be added within the user defined cost functions, but this allows a neater method for repeated use.

I wasn't sure if we wanted to do the aggregation of sea-states ourselves or leave it up to the user. It seemed like something the user might want to do themselves (particularly if they are creating bespoke outputs) so I added another callback to DefaultBlueprint called `aggregationHook` where you can do this (the result of the callback are added to the Device object's `aggregation` property). You'll see an example of this in the RM3 class and I've updated the example (not optimisation) to work with multiple sea-states. 

Let me know how this fits.

H0R5E commented:
@ryancoe, I'm going to keep going with this until I've got most of the engineering work done and then I will need to take a chunk of time for my other projects. I just think it's advantageous for this to not linger too long, half finished.

Also, we need to have a discussion about how we introduce this into the code base proper. I think it should be sooner, rather than later, even if unfinished, otherwise we are going to bottleneck all of the other PRs. Let's make it a priority to discuss on Thursday.

ryancoe commented:
One minor note: I think you may need to use some different keywords in your docstrings to help MATLAB recognize things.

https://www.mathworks.com/help/matlab/matlab_prog/create-help-for-classes.html

H0R5E commented:
> One minor note: I think you may need to use some different keywords in your docstrings to help MATLAB recognize things.
> 
> https://www.mathworks.com/help/matlab/matlab_prog/create-help-for-classes.html

Is there a specific example you can point to? There is a conflict between what Sphinx autodoc can read and what MATLAB can read. It might be a choice of favouring online docs or MATLAB docs.

H0R5E commented:
@ryancoe, quick question while I think about it. With the hydrodynamics calls that use NEMOH, a set of frequencies are required and we tend to set these from the first sea state of interest. Currently, I am passing this sea state as a fixed parameter to the blueprint, NEMOH is then called when the device is made, and then any other sea state can be used with the device to calculate the performance (by interpolating). 

I am wondering, though, if we should be calling NEMOH every time we use a new sea state or whether it should be run with a default set of values? This would either enforce or break this dependency on the initial sea state, which would be healthier, IMO.

Not sure if this is something we should deal with now, but I just wanted to pop it down, while it was at the top of thoughts.

ryancoe commented:
@H0R5E Good question. I think #143 will actually deal with this in a more robust way (using the same set of frequencies for all sea states and calling nemoh only once).

H0R5E commented:
Guys, while I was updating the tests I found that the master branch code has a bug in it for the determination of the A33 matrix of the RM3 device. In the master code it's given as:

```
B33 = squeeze(interp1(hydro.w,squeeze(hydro.B(3, 3, :)),w,'linear',0)) * hydro.rho .* w;
A33 = squeeze(interp1(hydro.w,squeeze(hydro.B(3, 3, :)),w,'linear',0)) * hydro.rho;
```

Note how A33 is found from **`hydro.B`**, which I think is incorrect and should be **`hydro.A`**. I would guess that any calculations using this algorithm are incorrect. It also shows that we could do with some better analytical tests so that we can catch this sort of typo in the future.

H0R5E commented:
Test results:

[test_results.pdf](https://github.com/SNL-WaterPower/WecOptTool/files/4736589/test_results.pdf)

H0R5E commented:
More tests...

[test_results.pdf](https://github.com/SNL-WaterPower/WecOptTool/files/4736675/test_results.pdf)

ssolson commented:
> Guys, while I was updating the tests I found that the master branch code has a bug in it for the determination of the A33 matrix of the RM3 device. In the master code it's given as:
> 
> ```
> B33 = squeeze(interp1(hydro.w,squeeze(hydro.B(3, 3, :)),w,'linear',0)) * hydro.rho .* w;
> A33 = squeeze(interp1(hydro.w,squeeze(hydro.B(3, 3, :)),w,'linear',0)) * hydro.rho;
> ```
> 
> Note how A33 is found from **`hydro.B`**, which I think is incorrect and should be **`hydro.A`**. I would guess that any calculations using this algorithm are incorrect. It also shows that we could do with some better analytical tests so that we can catch this sort of typo in the future.

The nearby code suggests is should be hydrp.A. @gbacelli can you confirm that this should be 
` A33 = squeeze(interp1(hydro.w,squeeze(hydro.A(3, 3, :)),w,'linear',0)) * hydro.rho;`

H0R5E commented:
@ryancoe and @ssolson, everything is done baring the online documentation. I started tidying it up by renaming the example as _"Optimizing an Existing WEC Model"_ and then adding another page called _"Creating a New WEC Model"_. My thought is that the "optimizing" page can be adapted from the old example page, but we will have to write something new for the "new model" page. We could always just put a placeholder if there for now, also.

I need to take a little time off this, to give my brain a break, so if you feel inclined to help, you can go ahead without conflicting with me. Any help you can give is gratefully received.

H0R5E commented:
OK, I'm going to start working on this again.

H0R5E commented:
I am done with the rewrite of the old example. @ryancoe how do we feel about the "Creating a New WEC Model" page? Is it OK to leave it for now and make an issue to complete it or do you want to me to do for this PR?

ryancoe commented:
I’d say ok to leave for now, but definitely add to our issues list.

H0R5E commented:
Cool, can you have a look at the example, then, when you get a moment? I'd say if you're happy with that, then we should go ahead and merge it. Work for you?

ryancoe commented on docs/user/optimization.rst:
```diff
@@ -0,0 +1,444 @@
+.. image:: /_static/WecOptTool_algorithmDiagram.svg
+   :alt: Conceptual illustration of WecOptTool functionality
+
+The WecOptTool framework encapsulates the above process using two key classes. 
+The :mat:class:`~+WecOptTool.Blueprint` class is used to describe the potential 
+forms (e.g. shape, control type, etc.) that a WEC device might take. The user 
+of WecOptTool must create a `concrete subclass 
+<https://uk.mathworks.com/help/matlab/matlab_oop/abstract-classes-and-interface 
+s.html>`_ of the :mat:class:`~+WecOptTool.Blueprint` class that describes the 
+potential geometry, motion and control options for their chosen WEC design. 
+For this example, the subclass is already prepared in the |RM3.m|_ file and 
+details regarding how this file was created is available in the :ref:`model` 
+page. 
+
+The :mat:class:`~+WecOptTool.Device` class represents a device with a chosen 
+set of design parameters, created from a Blueprint subclass. Subsequently, the 
+performance of Device objects can be tested against any seastate of choice. 
+In summary, the general process is:
+
+#. Create a Blueprint subclass object
+#. Create one or many Device objects using the Blueprint
+#. Examine the performance each Device object for a given seastate
+
+The remainder of this page will illustrate (using the |optimization.m|_ 
```
To avoid overwhelming a new user (potentially unfamiliar with many of these OOP terms), I'd suggest we move this content to `model.rst`.

H0R5E replied:
Agreed, this is too technical. I will water it down a bit.

ryancoe replied:
I think it is great, and we will need more stuff like this on that subject, but let's have something brief and unintimidating in this most basic tutorial.

ryancoe commented:
@H0R5E - It looks like you have completed most (or all) of the outstanding to-do's. Is there anything you need help with (or just final review)?

H0R5E commented:
> @H0R5E - It looks like you have completed most (or all) of the outstanding to-do's. Is there anything you need help with (or just final review)?

Yeah, everything is done, and I was thinking that going through the example would provide a sufficient "review" of the code functionality that is here. Because the thing is so big, it's probably better to ask for forgiveness on any problems with this one and merge it now, so that I can move onto reworking the other PRs.
